Louis/podcidrs bgp

ldelossa opened the pull request:
This pull adds the necessary commits to support BGP route withdrawal on a Node Delete event. 

After a Node Delete event is encountered an empty Advertisement will be plumbed to the BGP controller running inside the Cilium agent. 

After the event is received by our harness, we will flip a atomic to then shut the speaker down. When the BGP speaker is shut down it will no longer ingest concurrent events nor process any events that are still in its event queue.

ldelossa committed:
bgp: plumb context through speaker

this commit adds basic context handling in the
speaker.

constructors of a speaker should now cancel the
passed context to stop bgp events.

Signed-off-by: ldelossa <louis.delos@isovalent.com>

ldelossa committed:
bgp: implement withdrawal

this commit implements a concurrency safe bgp withdrawal mechanism which
is triggered via Node Delete events.

when a node delete event is encountered we now signal to the BGP
controller that it should withdrawal all previously advertised routes.

Signed-off-by: ldelossa <louis.delos@isovalent.com>

christarazi commented on pkg/bgp/speaker/events.go:
```diff
@@ -35,6 +36,10 @@ type nodeEvent struct {
+	// withDraw will be set when a Delete node event occurs.
+	// the reduction of this event will elicit a withdrawal
+	// of all bgp routes.
+	withDraw *bool
```
This doesn't need to be a pointer because `bool` is hashable. Making these fields pointers is only required for types that are not hashable such as maps or slices. 

I guess 
>// The following fields must be a pointers because they are not hashable
	// (read comparable) in Go.

was a bit misleading since it can be interpreted as all fields? Probably worth clarifying it.

ldelossa replied:
You are correct. 

I'm not exactly aware of why this consideration is here in the first place? I didn't see any code doing comparisons of these event types and fields. Can you clue me in?

christarazi replied:
The issue is that the event queuing uses a `map` under the hood, I believe. So if you were to remove all the pointers from the `nodeEvent`, you'll get a panic saying "unhashable type" or something like that when a `nodeEvent` comes in. https://blog.golang.org/maps#TOC_5.

ldelossa replied:
got it. It's required by the k8s workqueue.

christarazi commented on pkg/bgp/speaker/events.go:
```diff
@@ -46,7 +51,13 @@ func (s *Speaker) run(ctx context.Context) {
 		// only check ctx here, we'll allow any in-flight
 		// events to be processed completely.
-		if ctx.Err != nil {
+		if ctx.Err() != nil {
```
This change was probably supposed to be in the previous commit?

christarazi commented on pkg/bgp/speaker/pod_cidr.go:
```diff
@@ -25,6 +26,29 @@ import (
+			// advertised bgp routes.
+			err := session.Set(emptyAdverts...)
+			if err != nil {
+				log.Error("failed to withdrawal bgp routes")
```
`log.WithError(err) ...`

christarazi commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -170,6 +192,11 @@ func (s *Speaker) OnUpdateEndpoints(eps *slim_corev1.Endpoints) {
 	log.Infof("chris Speaker OnAddNode %v", node.GetName())
+
+	if s.shutDown() {
+		return fmt.Errorf("cannot enqueue event, speaker is shutdown")
```
Probably worth extracting out this error into its own variable via `errors.New()`

```
var (
  errShuttingDown = errors.New(...)
)
```

ldelossa replied:
I like that :+1:

christarazi commented on pkg/bgp/speaker/events.go:
```diff
@@ -91,6 +102,16 @@ func (s *Speaker) handleNodeEvent(k nodeEvent) types.SyncState {
+		// this is a best effort method call, so we don't
+		// care about errors. If the node is shutting down
+		// all routes will be closed once the speaker's TCP conn
+		// is closed anyway.
```
Would be nice to reference the RFC here

ldelossa replied:
agreed.

christarazi commented on pkg/bgp/speaker/pod_cidr.go:
```diff
@@ -25,6 +26,29 @@ import (
+			// advertised bgp routes.
+			err := session.Set(emptyAdverts...)
+			if err != nil {
+				log.WithError(err).Error("failed to withdrawal bgp routes")
```
```suggestion
				log.WithError(err).Error("Failed to withdraw BGP routes")
```

christarazi commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -181,7 +223,27 @@ func (s *Speaker) OnUpdateNode(oldNode, newNode *slim_corev1.Node) error {
+//
+// When the speaker discovers the node that it is running on
+// is shuttig down it will send a BGP message to its peer
+// instructing it to withdrawal all previously advertised
```
```suggestion
// instructing it to withdraw all previously advertised
```

christarazi commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -121,6 +142,9 @@ func (s *Speaker) OnUpdateService(svc *slim_corev1.Service) {
 
+	if s.shutDown() {
+		return
+	}
```
I thought that I refactored the other subscriber implementations to return errors as well, but I'm only seeing the node subscriber methods returning errors? If the others haven't been implemented, I think it's worth doing a refactor commit to do that. Doesn't have to be part of this specific PR though.

ldelossa replied:
I'll do this in this PR.

christarazi commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -34,12 +36,16 @@ import (
 )
 
+var (
+	ErrShutDown = errors.New("cannot enqueue event, speaker is shutdown")
```
Does this need to be exported?

ldelossa replied:
It would allow the k8s watcher to determine if it wants to react to the bgp speaker being shut down.

christarazi reviewed (approved):
Minor nits

ldelossa committed:
bgp: refactor method set to return errs

this commit refactors the bgp speaker method set to return errors when
k8s events cannot be processed.

Signed-off-by: ldelossa <louis.delos@isovalent.com>
